doc/DIALER.md

```diff
@@ -0,0 +1,36 @@
+# js-libp2p Dialer
+
+**Synopsis**
+* All Dial Requests in js-libp2p must request a token(s) from the Dialer.
+  * The number of tokens requested should be between 1 and the PER_PEER_LIMIT max set in the Dialer.
+  * If the number of available tokens is less than requested, the Dialer may return less than requested.
+* The number of tokens a DialRequest obtains reflects the maximum number of parallel Multiaddr Dials it can make.
+* If no tokens are available a DialRequest should immediately end and throw. This deviates from the existing queue system to avoid queue congestion and provide more visibility to users.
+* As tokens are limited, DialRequests should be given a prioritized list of Multiaddrs to minimize the potential request time.
+* Once a single Multiaddr Dial has succeeded, all pending dials in that Dial Request should be aborted. All tokens should be immediately released to the Dialer.
+* If all Multiaddr Dials fail, or the DIAL_TIMEOUT max is reached for the entire DialRequest, all in progress dials for that DialRequest should be aborted. All tokens should immediately be released to the Dialer.
```

Comment:
Should we try to release tokens as we don't need them?

Example: DialRequest gets 3 tokens and the peer has 3 multiaddr. As soon as the first dial attempt fails (from the 3 parallel requests), can we immediatly release that resource?

Reply:
Yes we should definitely do this. I will update to reflect this.